aten/src/ATen/native/MaxUnpooling.cpp

```diff
@@ -0,0 +1,588 @@
+#include <tuple>
+
+namespace at {
+namespace native {
+
+template <typename scalar_t>
+Tensor max_unpooling2d_forward_out_cpu_frame(
+    Tensor& output,
+    const Tensor& input,
+    const Tensor& indices,
+    int64_t outputHeight,
+    int64_t outputWidth) {
+  int64_t numBatch = 1;
+  int64_t dimc = 0;
+  int64_t dimh = 1;
+  int64_t dimw = 2;
+  if (input.ndimension() == 4) {
+    numBatch = input.size(0);
+    dimc++;
+    dimh++;
+    dimw++;
+  }
+  int64_t numChannels = input.size(dimc);
+  int64_t inputHeight = input.size(dimh);
+  int64_t inputWidth = input.size(dimw);
+
+  auto* rawInput = input.data<scalar_t>();
+  auto* rawIndices = indices.data<int64_t>();
+  auto* rawOutput = output.data<scalar_t>();
+
+  for (int64_t n = 0; n < numBatch; n++) {
+    int64_t nOutputOffset = n * numChannels * outputWidth * outputHeight;
+    int64_t nInputOffset = n * numChannels * inputWidth * inputHeight;
+    int64_t k;
+    bool has_error = false;
+    int64_t error_index = 0;
+#pragma omp parallel for private(k)
+    for (k = 0; k < numChannels; k++) {
+      int64_t finalOutputOffset =
+          nOutputOffset + k * outputWidth * outputHeight;
+      int64_t finalInputOffset = nInputOffset + k * inputWidth * inputHeight;
+      scalar_t* output_p_k = rawOutput + finalOutputOffset;
+      scalar_t* input_p_k = rawInput + finalInputOffset;
+      int64_t* ind_p_k = rawIndices + finalInputOffset;
+
+      int64_t maxp;
+      for (int64_t i = 0; i < inputHeight; i++) {
+        for (int64_t j = 0; j < inputWidth; j++) {
+          maxp = ind_p_k[i * inputWidth + j];
+          if (maxp < 0 || maxp >= outputWidth * outputHeight) {
+#pragma omp critical
+            {
+              has_error = true;
+              error_index = maxp;
+            }
+          } else {
+            output_p_k[maxp] = input_p_k[i * inputWidth + j];
+          }
+        }
+      }
+    }
+    if (has_error) {
+      AT_ERROR(
+          "Found an invalid max index",
+          error_index,
+          " (output volumes are of size ",
+          outputHeight,
+          "x",
+          outputWidth);
+    }
+  }
+  return output;
+}
+
+Tensor& max_unpooling2d_forward_out_cpu(
+    Tensor& output,
+    const Tensor& self,
+    const Tensor& indices,
+    IntList output_size) {
+  AT_CHECK(
+      indices.scalar_type() == at::ScalarType::Long,
+      "elements in indices should be type Long");
+  AT_CHECK(
+      output_size.size() == 2,
+      "There should be exactly two elements (height, width) in output_size");
+  AT_CHECK(
+      (self.ndimension() == 3 || self.ndimension() == 4),
+      "Input to max_unpooling2d should be a 3d or 4d Tensor");
+  AT_CHECK(
+      self.sizes() == indices.sizes(),
+      "Shape of indices should match shape of input");
+
+  AT_CHECK(self.numel() > 0, "Input must be non-empty");
+
+  auto outputHeight = output_size[0];
+  auto outputWidth = output_size[1];
+
+  auto self_contiguous = self.contiguous();
```

Comment:
nit: it's really easy to forget to use `self_contiguous` rather than `self` in the code below, so I like to name the non-contiguous variable `self_` (in the argument list) and the contiguous variable `self` so it's hard to get wrong.

Reply:
(And in fact, you are mixing uses of `self` and `self_contiguous` in the code below. Functionally it doesn't matter, but it makes it more easy to make a mistake and use `self` rather than `self_contiguous` when using the contiguous version is mandatory. Why make it possible to make this mistake?)